portfolio/src/main/java/com/google/sps/servlets/SpotifyServlet.java

```diff
@@ -0,0 +1,150 @@
+
+  private static final String clientId = "444dda318d8949cb9282d95a9885a7c6";
+  private static final String clientSecret = "9f81b8a6cf5748ef9d24a7ddb1e8bc4c";
+  private static final String redirectUriString = "https://8080-b0939ef0-d7b7-4da5-abf9-d0bf54e01dcf.us-east1.cloudshell.dev/spotify";
+  private static final URI redirectUri = SpotifyHttpManager.makeUri(redirectUriString);
+  private static String oauthToken = "";
+  
+  private static final SpotifyApi spotifyApi = new SpotifyApi.Builder()
+    .setClientId(clientId)
+    .setClientSecret(clientSecret)
+    .setRedirectUri(redirectUri)
+    .build();
+  
+  private static final AuthorizationCodeUriRequest authorizationCodeUriRequest = spotifyApi.authorizationCodeUri()
+    .scope("user-read-currently-playing")
+    .show_dialog(true)
+    .build();
+ 
+  private static AuthorizationCodeRequest authorizationCodeRequest = spotifyApi.authorizationCode(oauthToken)
+    .build();
+
+  private static GetUsersCurrentlyPlayingTrackRequest getUsersCurrentlyPlayingTrackRequest = spotifyApi
+    .getUsersCurrentlyPlayingTrack()
+    .build();
+
+  private static AuthorizationCodeRefreshRequest authorizationCodeRefreshRequest = spotifyApi.authorizationCodeRefresh()
+    .build();
+
+  public static String authorizationCodeUri() {
+    final URI uri = authorizationCodeUriRequest.execute();
+    return uri.toString();
+  }
+
+  public static String formatAsMinutesAndSeconds(Integer milliseconds) {
+    Integer seconds = (Integer)(milliseconds/ 1000) % 60;
+    Integer minutes = (Integer)(milliseconds/ (1000) / 60);
+    return String.format("%d:%02d", minutes, seconds);
+  }
+
+  public static void authorizationCode() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRequest.execute();
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static void authorizationCodeRefresh() {
+    try {
+      final AuthorizationCodeCredentials authorizationCodeCredentials = authorizationCodeRefreshRequest.execute();
+ 
+      spotifyApi.setAccessToken(authorizationCodeCredentials.getAccessToken());
+      spotifyApi.setRefreshToken(authorizationCodeCredentials.getRefreshToken());
+ 
+      System.out.println("Expires in: " + authorizationCodeCredentials.getExpiresIn());
+    } catch (IOException | SpotifyWebApiException | ParseException e) {
+      System.out.println("Error: " + e.getMessage());
+    }
+  }
+
+  public static String getCurrentTrack() {
+    try {
+      final CompletableFuture<CurrentlyPlaying> currentlyPlayingFuture = getUsersCurrentlyPlayingTrackRequest.executeAsync();
```

Comment:
Is there a non-async version of this function (.execute())? If you 'join' the future immediately after creating it you're not actually getting any benefit from being async

Reply:
I see what you mean, and yes there is a non-async version. I'll use that version instead!